docs/tutorials/flyscanning/plotting.md

```diff
@@ -0,0 +1,212 @@
+---
+jupytext:
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+```{code-cell} ipython3
+
+```
+
+# Live Plotting of Streamed Data
+
+A distictive characteristic of fly scanning plans is their reliance on continuous data streams rather than acquisition of data points invidually. In Bluesky, data streaming is realized via StreamResource and StreamDatum documents (as aooposed to Resource and Datum in step scanning). The eventual number of data points in the stream (e.g. image frames acquired during a flyscanning experiment) is typically unknown at the begininng of the aquisition; furtehrmore, the frames may arrive at non-uniformly spaced time intervals. Eventually, the frames become parts of the same resource and are often appended to the same file (in case of hdf5) by the detector.
+
+
+We introduce the notion of Consolidators to facilitate random access by data access services, e.g. Tiled. Consolidators are analogous to Handlers that have existed in Bluesky for a long time, but they enable working with streamed data. A Consolidator is defined for a particular stream of data and is tied to the underlying resource; it knows how to read the data from disk in the most efficient manner.
+
+
+In this notebook, we will explore the use of Consolidators as an option for live plotting of streamed data.
+
+```{code-cell} ipython3
+%load_ext autoreload
+%autoreload 2
+
+import os
+import asyncio
+import ophyd_async
+import bluesky
+from ophyd_async.sim.demo import PatternDetector, SimMotor
+import bluesky.plans as bp
+import bluesky.plan_stubs as bps
+import bluesky.preprocessors as bpp
+from bluesky.callbacks.core import CollectLiveStream
+from bluesky.callbacks.mpl_plotting import LiveStreamPlot
+from bluesky.run_engine import RunEngine
+from pathlib import Path
+from sample_simulator import SampleSimulator
+from json_writer import JSONWriter
+
+
+print(f"{ophyd_async.__version__=}\n{bluesky.__version__    =}")
+```
+
+```{code-cell} ipython3
+import matplotlib
+import matplotlib.pyplot as plt
+
+%matplotlib widget
+plt.ion()
+```
+
+```{code-cell} ipython3
+fig, ax = plt.subplots(1, 2, figsize=(13, 4))
+```
+
+```{code-cell} ipython3
+det = PatternDetector(name="PATTERN1", path=Path('.'))
+motx = SimMotor(name="Motor_X", instant=False)
+moty = SimMotor(name="Motor_Y", instant=False)
+await motx.velocity.set(2.0)
+await moty.velocity.set(1.0)
+
+RE = RunEngine({}, during_task = SampleSimulator(det, motx, moty))
```

Comment:
Interesting! This hook is meant to run things like plotting event loops. I'm not sure how I feel about using it to run a simulator like this. Maybe? Is there precedent elsewhere?

Reply:
No, I don't think there's a precedent, or at least I haven't seen one.
Yes, understand that the hook was meant for plotting, but this was the best I could think of after many attempts to connect a simulated motor to a simulated detector. The detector has positional attributes, x and y, that determine the produced signal; as I understood, they are meant to represent the position of the motor, but this connection should be realized outside and independent of the bluesky plan (i.e. to simulated an actual physical sample). Perhaps, there's a better way, but we'll likely need to expand the simulated detector and/or motor classes.

Reply:
This is an important general question. Sometimes we have run simulations like this as separate processes, sometimes as IOCs. Sometimes we have used threads, as a couple objects in `ophyd.sim` do. For the cookbook in general, we should think through the trade-offs and perhaps document a recommended way to drive simulations external to the Bluesky plan.

Reply:
As part of the ophyd-async 1.0 release I'll make something that looks very much like this in `ophyd_async.sim`